feat: support customize `Header` and improve validations

greenhat616 opened the pull request:
* Use `clap` to parse `env` directly
* Support customize `headers` for SSE/StreamableHttp transport
* Support specify `TransportType` directly
* validate `Url` while parsing `Args`

greenhat616 committed:
feat: support connection headers

greenhat616 committed:
feat: add url validation

greenhat616 committed:
fix: make headers work

fix: make headers work

josevalim commented:
Thank you @greenhat616! Can you please explain your use cases? The reason we ask is because we moved most of our servers to use streamable http and we hope we won't need the proxy for much longer, so we are worried about adding features to the project when we may not maintain it for much longer.

greenhat616 commented:
@josevalim We just migrated to streamable http too. To compatibale with `Claude Desktop`, we have to look for a stdio proxy to make it run on local mcp servers.

FYI, Claude Desktop only support SSE/Streamable HTTP MCP Server on `Settings / Connector`, which is only available for Pro/Max or Ent users. The local MCP Servers only support stdio servers.

<img width="484" height="810" alt="图片" src="https://github.com/user-attachments/assets/d7057be1-c837-4edc-bc89-e2a56a76842b" />


* https://thesof.medium.com/bridging-mcp-transports-turning-an-stdio-server-into-sse-and-the-other-way-around-da23a36e074a

nzjrs commented:
I ran into this limitation today too. I need arbitrary header support to bypass a proxy to access and internal resource, externally.

SteffenDE commented:
The code looks good to me, but just to align expectations: we maintain this proxy for the Tidewave MCP servers that are meant to run on the same machine. The features we have (replying init for reconnects) are very much focused on that specific use case. Adding features that are not needed for this are therefore not a priority for us and as José mentioned we may very well archive this repository once we deem it unnecessary for our use case.

If you want to provide a more generalized MCP proxy based on this project, you're more than welcome to do so in a fork as we're unlikely to accept many new features here.

greenhat616 committed:
feat: migrate to rmcp 1.6.0 from forked dependency

- Replace SteffenDE/mcp-rust-sdk fork with official rmcp 1.6.0
- Upgrade reqwest from 0.12 to 0.13
- Remove SseClientTransport and SseServer (removed in rmcp 1.6)
- Unify transport to single StreamableHttpClientTransport
- Remove SseClientType enum, add ServerTransport type alias
- Deprecate --transport-type=sse CLI option (warns + falls back)
- Rename sse_url to server_url (keep SSE_URL env for compat)
- Use StreamableHttpClientTransportConfig builder pattern
- Adapt to #[non_exhaustive] structs (ServerInfo, CallToolRequestParams)
- Update examples and tests to use StreamableHttpService
- Add no_proxy() to reqwest client to avoid proxy interference

Note: 3 advanced tests (reconnection, ping-when-disconnected,
protocol-version-override) need further work to adapt to
StreamableHttpClientTransport's behavioral differences
(no persistent connection for disconnect detection, session
keepalive timeout affects port reuse in tests).

greenhat616 committed:
fix: only disable proxy when NO_PROXY env var is set

Previously reqwest client always had no_proxy() applied, which
prevented legitimate proxy configurations from working. Now the
no_proxy setting is conditional on the NO_PROXY/no_proxy env var.

greenhat616 committed:
Fix proxy tests on Windows
